lib/datadog/profiling/stack_recorder.rb

```diff
@@ -5,8 +5,12 @@ module Profiling
     # Used to wrap a ddprof_ffi_Profile in a Ruby object and expose Ruby-level serialization APIs
     # Methods prefixed with _native_ are implemented in `stack_recorder.c`
     class StackRecorder
+      def initialize
+        @no_concurrent_synchronize_mutex = Thread::Mutex.new
+      end
+
       def serialize
-        status, result = self.class._native_serialize(self)
+        status, result = @no_concurrent_synchronize_mutex.synchronize { self.class._native_serialize(self) }
```

Comment:
What is `@no_concurrent_synchronize_mutex` guarding? I don't think I understand the reason for this specific mutex.

Reply:
Added in 6fd351fd9667907859f7915e8bc65de8ff5d8fff the following note to the code:

This mutex works in addition to the fancy C-level mutexes we have in the native side (see the docs there). It prevents multiple Ruby threads calling serialize at the same time -- something like `10.times { Thread.new { stack_recorder.serialize } }`.

This isn't something we expect to happen normally, but because it would break the assumptions of the C-level mutexes (that there is a single serializer thread), we add it here as an extra safeguard against it accidentally happening.

(Note that this usually isn't a problem when calling from Ruby into native because the native side continues to hold on to the global VM lock. But since `serialize` releases the GVL, it's possible that a second Ruby thread would come in an execute it again.)

Reply:
I'm happy to keep `@no_concurrent_synchronize_mutex`, then thank you for the explanation.

I'm curious: if someone does `10.times { Thread.new { stack_recorder.serialize } }` without the presence of `@no_concurrent_synchronize_mutex`, what happens? If our native code simply returns a nice Ruby error, I don't mind not having the `@no_concurrent_synchronize_mutex`. But if it causes weird lock states, or undefined behavior then the mutex is surely a must-have. Regardless, I see no harm in keeping it.